core/node/vm_runner/src/output_handler.rs

```diff
@@ -0,0 +1,541 @@
+use std::{
+    fmt::{Debug, Formatter},
+    sync::Arc,
+    time::Duration,
+};
+
+use async_trait::async_trait;
+use dashmap::DashMap;
+use futures::future::BoxFuture;
+use tokio::sync::{oneshot, watch};
+use zksync_core::state_keeper::{StateKeeperOutputHandler, UpdatesManager};
+use zksync_dal::{ConnectionPool, Core};
+use zksync_types::L1BatchNumber;
+
+use crate::VmRunnerIo;
+
+/// Functionality to produce a [`StateKeeperOutputHandler`] implementation for a specific L1 batch.
+///
+/// The idea behind this trait is that often handling output data is independent of the order of the
+/// batch that data belongs to. In other words, one could be handling output of batch #100 and #1000
+/// simultaneously. Implementing this trait signifies that this property is held for the data the
+/// implementation is responsible for.
+#[async_trait]
+pub trait OutputHandlerFactory: Debug + Send {
+    /// Creates a [`StateKeeperOutputHandler`] implementation for the provided L1 batch. Only
+    /// supposed to be used for the L1 batch data it was created against. Using it for anything else
+    /// is undefined behavior.
```

Comment:
UB has specific meaning in Rust 🙃 AFAIU, in this case the use will lead to an error, which is much better than UB (which could e.g. mean a segfault).